core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -184,6 +246,41 @@ class AdminClient(val time: Time,
 
 }
 
+class CompositeFuture[T](client: ConsumerNetworkClient,
+                         time: Time,
+                         defaultResults: Map[TopicPartition, T],
+                         futures: List[RequestFuture[Map[TopicPartition, T]]]) extends Future[Map[TopicPartition, T]] {
+
+  override def isCancelled = false
+
+  override def cancel(interrupt: Boolean) = false
+
+  override def get(): Map[TopicPartition, T] = {
+    get(Long.MaxValue, TimeUnit.MILLISECONDS)
+  }
+
+  override def get(timeout: Long, unit: TimeUnit): Map[TopicPartition, T] = {
+    val start: Long = time.milliseconds()
+    val timeoutMs = unit.toMillis(timeout)
+    var remaining: Long = timeoutMs
+
+    val observedResults = futures.flatMap{ future =>
+      val elapsed = time.milliseconds() - start
+      remaining = timeoutMs - elapsed
+      if (remaining > 0 && client.poll(future, remaining)) future.value()
```

Comment:
This should not be a problem for `purgeDataBefore()`, because those `futures` provided to the `CompositeFuture` has been constructed in such a way that they never raises exception. Those future will call `future.complete(result)` in case of `onFailure`, where result indicates has the error information.

I agree it would be make  `CompositeFuture` more useful if this class handles the logic of converting error to result and return the full results to user as you suggested. But I don't have a good way to do it now because `CompositeFuture` doesn't know the type of the return value -- it currently use template `T`.